[release-4.15] OCPBUGS-36264: add Proxy config

LiangquanLi930 opened the pull request:
manual cherry-pick https://github.com/openshift/origin/pull/28880

LiangquanLi930 commented:
/jira refresh

LiangquanLi930 commented:
/jira refresh

LiangquanLi930 commented:
/retest-required

lihongan commented:
/label qe-approved

LiangquanLi930 commented:
/test e2e-gcp-ovn

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
@frobware Could you help review this PR? Thanks

LiangquanLi930 commented:
/cherry-pick release-4.14

Miciah commented on test/extended/router/http2.go:
```diff
@@ -48,6 +48,7 @@ func makeHTTPClient(useHTTP2Transport bool, timeout time.Duration) *http.Client
 		Timeout: timeout,
 		Transport: &http.Transport{
 			TLSClientConfig: &tlsConfig,
+			Proxy:           http.ProxyFromEnvironment,
```
This transport is overridden if `useHTTP2Transport` is true, so it seems like the test will still fail for the `useHTTP2Transport: true` test cases if using a proxy is required.  

Would it make sense to backport #28540 too?  This is the crucial change from #28540 to avoid overriding the transport that has the proxy configuration:
```diff
 	if useHTTP2Transport {
-		c.Transport = &http2.Transport{
-			TLSClientConfig: &tlsConfig,
-		}
+		// Explicitly allow HTTP/2.
+		http2.ConfigureTransport(transport)
 	}
```

candita replied:
@Miciah are you available to review this?  I thought we might have to expedite a change to the OWNERS file.

Miciah replied:
I'll make a best effort.  However, updating `OWNERS` shouldn't be blocking anyone from reviewing PRs; if someone else from the team has capacity to review, then that person can add an `/lgtm` and then ping me, Miheer, or knobunc to add a rubber-stamp `/lgtm`.

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

candita commented:
/retitle [release-4.15] OCPBUGS-36264: add Proxy config

Miciah commented:
@LiangquanLi930, thank you for following up https://github.com/openshift/origin/pull/28912#discussion_r1671675800!  However, the current PR is a little confusing.  It seems that you cherry-picked multiple commits, made some additional nontrivial changes (perhaps as part of conflict resolution), and squashed everything into a single commit in a way that removed authorship and makes it difficult to map the changes to the original commits.  

From what I can tell, you pulled in the following changes:

- https://github.com/openshift/origin/pull/28514/commits/cdb5cd60c6188d7e890988968e363b23589ad7af – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
- https://github.com/openshift/origin/pull/28514/commits/ef8cb92fa83ac895b8931173b6f412c4f38231bb – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
- https://github.com/openshift/origin/pull/28522/commits/a2013b71b9702f5705943b63b776f65eb0f2cdce – However, you kept `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }` even though the original commit deleted it.  Presumably you decided it didn't hurt anything to keep it, but it would be useful to note this decision in the commit message so we can be sure it wasn't dropped by accident during conflict resolution.
- https://github.com/openshift/origin/pull/28540/commits/b9fac649dca12bf52c393ba6a16ed99a7056acc8 – This change looks fine.  
- https://github.com/openshift/origin/pull/28880/commits/55255b0675b891693702ac72a1d50e3b4ed9fb55 – This is the original change you wanted to backport.  

Is that a complete and accurate list?

For completeness, I'll note that the following changes to `http2.go` and `h2spec.go` from the main branch were not pulled in:

- https://github.com/openshift/origin/pull/28515/commits/60f67f2e0381bf305a05b8f17de20aa346d9798c – Omitting this change is fine as it was reverted on the main branch.
- https://github.com/openshift/origin/pull/28515/commits/00ea63b861570609c0bd7b02254c303772ea5b33 – Ditto.
- https://github.com/openshift/origin/pull/28757/commits/1f14f779539eda5096a677834914ea6876f8ca18 – Omitting this change is fine as it is an unrelated bug fix.  (If anyone requests a backport, we may need to deal with merge conflicts then, but so far no one has asked for a backport.)

Usually, I would expect to see each cherry-picked commit preserved in the backport PR (amended as necessary to resolve merge conflicts or possibly to make other necessary changes).  If you need to squash the commits (I know some teams do this for reasons that I don't understand), then please at least make sure that the commit message lists the cherry-picked commits and mentions your changes (in particular, please mention that you restored `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }`, and why).  And at the _very_ least, please make sure that you preserve authorship.

LiangquanLi930 commented:
> @LiangquanLi930, thank you for following up [#28912 (comment)](https://github.com/openshift/origin/pull/28912#discussion_r1671675800)! However, the current PR is a little confusing. It seems that you cherry-picked multiple commits, made some additional nontrivial changes (perhaps as part of conflict resolution), and squashed everything into a single commit in a way that removed authorship and makes it difficult to map the changes to the original commits.
> 
> From what I can tell, you pulled in the following changes:
> 
> * [cdb5cd6](https://github.com/openshift/origin/commit/cdb5cd60c6188d7e890988968e363b23589ad7af) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> * [ef8cb92](https://github.com/openshift/origin/commit/ef8cb92fa83ac895b8931173b6f412c4f38231bb) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> * [a2013b7](https://github.com/openshift/origin/commit/a2013b71b9702f5705943b63b776f65eb0f2cdce) – However, you kept `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }` even though the original commit deleted it.  Presumably you decided it didn't hurt anything to keep it, but it would be useful to note this decision in the commit message so we can be sure it wasn't dropped by accident during conflict resolution.
> * [b9fac64](https://github.com/openshift/origin/commit/b9fac649dca12bf52c393ba6a16ed99a7056acc8) – This change looks fine.
> * [55255b0](https://github.com/openshift/origin/commit/55255b0675b891693702ac72a1d50e3b4ed9fb55) – This is the original change you wanted to backport.
> 
> Is that a complete and accurate list?
> 
> For completeness, I'll note that the following changes to `http2.go` and `h2spec.go` from the main branch were not pulled in:
> 
> * [60f67f2](https://github.com/openshift/origin/commit/60f67f2e0381bf305a05b8f17de20aa346d9798c) – Omitting this change is fine as it was reverted on the main branch.
> * [00ea63b](https://github.com/openshift/origin/commit/00ea63b861570609c0bd7b02254c303772ea5b33) – Ditto.
> * [1f14f77](https://github.com/openshift/origin/commit/1f14f779539eda5096a677834914ea6876f8ca18) – Omitting this change is fine as it is an unrelated bug fix.  (If anyone requests a backport, we may need to deal with merge conflicts then, but so far no one has asked for a backport.)
> 
> Usually, I would expect to see each cherry-picked commit preserved in the backport PR (amended as necessary to resolve merge conflicts or possibly to make other necessary changes). If you need to squash the commits (I know some teams do this for reasons that I don't understand), then please at least make sure that the commit message lists the cherry-picked commits and mentions your changes (in particular, please mention that you restored `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }`, and why). And at the _very_ least, please make sure that you preserve authorship.

@Miciah Thanks for your help, follow your comments, I updated the pr to keep the original back-port commits and then add my back-port pr on the latest commit

candita commented:
/assign
/assign @gcs278

Miciah commented:
> > @LiangquanLi930, thank you for following up [#28912 (comment)](https://github.com/openshift/origin/pull/28912#discussion_r1671675800)! However, the current PR is a little confusing. It seems that you cherry-picked multiple commits, made some additional nontrivial changes (perhaps as part of conflict resolution), and squashed everything into a single commit in a way that removed authorship and makes it difficult to map the changes to the original commits.
> > From what I can tell, you pulled in the following changes:
> > 
> > * [cdb5cd6](https://github.com/openshift/origin/commit/cdb5cd60c6188d7e890988968e363b23589ad7af) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> > * [ef8cb92](https://github.com/openshift/origin/commit/ef8cb92fa83ac895b8931173b6f412c4f38231bb) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> > * [a2013b7](https://github.com/openshift/origin/commit/a2013b71b9702f5705943b63b776f65eb0f2cdce) – However, you kept `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }` even though the original commit deleted it.  Presumably you decided it didn't hurt anything to keep it, but it would be useful to note this decision in the commit message so we can be sure it wasn't dropped by accident during conflict resolution.
> > * [b9fac64](https://github.com/openshift/origin/commit/b9fac649dca12bf52c393ba6a16ed99a7056acc8) – This change looks fine.
> > * [55255b0](https://github.com/openshift/origin/commit/55255b0675b891693702ac72a1d50e3b4ed9fb55) – This is the original change you wanted to backport.
> > 
> > Is that a complete and accurate list?
> > For completeness, I'll note that the following changes to `http2.go` and `h2spec.go` from the main branch were not pulled in:
> > 
> > * [60f67f2](https://github.com/openshift/origin/commit/60f67f2e0381bf305a05b8f17de20aa346d9798c) – Omitting this change is fine as it was reverted on the main branch.
> > * [00ea63b](https://github.com/openshift/origin/commit/00ea63b861570609c0bd7b02254c303772ea5b33) – Ditto.
> > * [1f14f77](https://github.com/openshift/origin/commit/1f14f779539eda5096a677834914ea6876f8ca18) – Omitting this change is fine as it is an unrelated bug fix.  (If anyone requests a backport, we may need to deal with merge conflicts then, but so far no one has asked for a backport.)
> > 
> > Usually, I would expect to see each cherry-picked commit preserved in the backport PR (amended as necessary to resolve merge conflicts or possibly to make other necessary changes). If you need to squash the commits (I know some teams do this for reasons that I don't understand), then please at least make sure that the commit message lists the cherry-picked commits and mentions your changes (in particular, please mention that you restored `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }`, and why). And at the _very_ least, please make sure that you preserve authorship.
> 
> @Miciah Thanks for your help, follow your comments, I updated the pr to keep the original back-port commits and then add my back-port pr on the latest commit

Sorry if I wasn't clear before.  I would prefer if this backport PR **exclude** these commits if they are not required:

- https://github.com/openshift/origin/commit/60f67f2e0381bf305a05b8f17de20aa346d9798c
- https://github.com/openshift/origin/commit/00ea63b861570609c0bd7b02254c303772ea5b33
- https://github.com/openshift/origin/commit/1f14f779539eda5096a677834914ea6876f8ca18

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
> > > @LiangquanLi930, thank you for following up [#28912 (comment)](https://github.com/openshift/origin/pull/28912#discussion_r1671675800)! However, the current PR is a little confusing. It seems that you cherry-picked multiple commits, made some additional nontrivial changes (perhaps as part of conflict resolution), and squashed everything into a single commit in a way that removed authorship and makes it difficult to map the changes to the original commits.
> > > From what I can tell, you pulled in the following changes:
> > > 
> > > * [cdb5cd6](https://github.com/openshift/origin/commit/cdb5cd60c6188d7e890988968e363b23589ad7af) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> > > * [ef8cb92](https://github.com/openshift/origin/commit/ef8cb92fa83ac895b8931173b6f412c4f38231bb) – This change should be fine for OpenShift 4.15 / HAProxy 2.6.
> > > * [a2013b7](https://github.com/openshift/origin/commit/a2013b71b9702f5705943b63b776f65eb0f2cdce) – However, you kept `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }` even though the original commit deleted it.  Presumably you decided it didn't hurt anything to keep it, but it would be useful to note this decision in the commit message so we can be sure it wasn't dropped by accident during conflict resolution.
> > > * [b9fac64](https://github.com/openshift/origin/commit/b9fac649dca12bf52c393ba6a16ed99a7056acc8) – This change looks fine.
> > > * [55255b0](https://github.com/openshift/origin/commit/55255b0675b891693702ac72a1d50e3b4ed9fb55) – This is the original change you wanted to backport.
> > > 
> > > Is that a complete and accurate list?
> > > For completeness, I'll note that the following changes to `http2.go` and `h2spec.go` from the main branch were not pulled in:
> > > 
> > > * [60f67f2](https://github.com/openshift/origin/commit/60f67f2e0381bf305a05b8f17de20aa346d9798c) – Omitting this change is fine as it was reverted on the main branch.
> > > * [00ea63b](https://github.com/openshift/origin/commit/00ea63b861570609c0bd7b02254c303772ea5b33) – Ditto.
> > > * [1f14f77](https://github.com/openshift/origin/commit/1f14f779539eda5096a677834914ea6876f8ca18) – Omitting this change is fine as it is an unrelated bug fix.  (If anyone requests a backport, we may need to deal with merge conflicts then, but so far no one has asked for a backport.)
> > > 
> > > Usually, I would expect to see each cherry-picked commit preserved in the backport PR (amended as necessary to resolve merge conflicts or possibly to make other necessary changes). If you need to squash the commits (I know some teams do this for reasons that I don't understand), then please at least make sure that the commit message lists the cherry-picked commits and mentions your changes (in particular, please mention that you restored `if tc.statusCode == 0 { if tc.statusCode == 0; return false, nil }`, and why). And at the _very_ least, please make sure that you preserve authorship.
> > 
> > 
> > @Miciah Thanks for your help, follow your comments, I updated the pr to keep the original back-port commits and then add my back-port pr on the latest commit
> 
> Sorry if I wasn't clear before. I would prefer that this backport PR **exclude** these commits if they are not required:
> 
> * [60f67f2](https://github.com/openshift/origin/commit/60f67f2e0381bf305a05b8f17de20aa346d9798c)
> * [00ea63b](https://github.com/openshift/origin/commit/00ea63b861570609c0bd7b02254c303772ea5b33)
> * [1f14f77](https://github.com/openshift/origin/commit/1f14f779539eda5096a677834914ea6876f8ca18)

Thank you very much for your patient explanation. I previously misunderstood, but I have now made the changes.

candita commented:
This was originally a simple one-line change that was required for proper test configuration of the private cluster for hypershift agent CI.  Upon backport it acquired a few more commits. Risk is low and should help CI instead of causing any issues.
/label backport-risk-assessed

All the commits are clean and ordered properly.  Still need to investigate the failing tests.
/lgtm
/approve

knobunc commented:
/approve

candita commented:
/test e2e-metal-ipi-ovn-ipv6

candita commented:
Let's see if the bug found and [fixed](https://github.com/openshift-metal3/dev-scripts/pull/1680) by @LiangquanLi930 is ready yet.
/test e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/retest

LiangquanLi930 commented:
/test e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
/test e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
/test e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
/test e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
Based on the log from [this link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/28912/pull-ci-openshift-origin-release-4.15-e2e-metal-ipi-ovn-ipv6/1818622917685022720), the cluster consistently fails to install, so I think we can ignore it.

LiangquanLi930 commented:
Based on past jobs, the job at https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-origin-release-4.15-e2e-metal-ipi-ovn-ipv6 will fail permanently, not just the current job.

Miciah commented:
/override ci/prow/e2e-metal-ipi-ovn-ipv6

knobunc commented:
/override ci/prow/e2e-metal-ipi-ovn-ipv6

LiangquanLi930 commented:
@sdodson Hi, Could you help to merge this PR? Thanks

LiangquanLi930 commented:
@deads2k @sdodson Hi, Could you help to merge this PR? Thanks

sdodson commented:
Got it, thanks @LiangquanLi930
